### ISSUE TITLE
Janus not respecting SIP Record-Route

### ISSUE THREAD

**Issue by s-schumann** [meetecho/janus-gateway#451]
Hi

Janus does not seem to respect Record-Route when sending in-dialog SIP messages.

We are using the SIP Webphone and found the following happen:
- Out of dialog INVITE/200 OK contain Record-Route headers
- ACK message builds Route headers accordingly and correctly
- ACK is sent directly to the other party (to Contact IP:Port) instead of the IP destination of the first Route 

Edit: If "proxy" is filled it works as expected, but I would interpret this as outbound proxy. If the proxy is found based on domain I would expect similar behavior.

Best regards
Sebastian

**Comment by saghul** [comment]
Hi Sebastian!

Generally speaking I found Sofia SIP acts quite weird: if you pass a non-resolved proxy, it will do the resolution *for every request*! So, I ended up making the resolution myself and passing the resulting IP address as the proxy.

**Comment by s-schumann** [comment]
Hi Saul
Thank you for replying.
For me, adding the proxy manually also worked. I've put the IP now and it is well usable :)
Strange though that Sofia SIP does not support Record-Route for my scenario w/o proxy, it isn't used that rare I think :(
Sebastian

**Comment by saghul** [comment]
Oh, it most certainly supports it. Alas the documentation is hard to follow, and the project is currently abandoned AFAIK.

**Comment by lminiero** [comment]
Do you guys think we should handle this "resolve the address" part in the plugin code, rather than forcing users to resolve externally themselves and then pass the proxy IP address?

**Comment by saghul** [comment]
Ideally yeah, I think so. I thought about using libunbound for it, but
never found the time to actually jump in and do it.

**Comment by lminiero** [comment]
Why another library instead of just using ```getaddrinfo```?

**Comment by saghul** [comment]
`getaddrinfo` doesn't do NAPTR nor SRV. We need a library capable of doing NAPTR and SRV lookups, and implement RFC 3263 with it.

**Comment by lminiero** [comment]
Got it, tnx for the clarification! I thought it did as there's a service as part of the arguments, but I never actually played with that so I was just guessing.

**Comment by lminiero** [comment]
Would c-ares work as well? Just wondering about which library is most likely to be already available, or more in general most likely to be available in distro repos.

**Comment by lminiero** [comment]
Answering myself after reading a bit on this, I guess you suggested libunbound for its ability to cache results which c-ares apparently doesn't have, am I right?

**Comment by saghul** [comment]
Both would work for this, but Unbound is made in The Netherlands, homeland of DNS, the Dutch Name System :-P

**Comment by saghul** [comment]
Also, the cache is nice :-)

**Comment by lminiero** [comment]
Just noticed we never got to add the resolving part... any easy to use examples/tutorials on libunbound out there? I'll try to look into this after the IETF, but I'd rather not study a cryptic or undocumented API, and I remember at the time I couldn't find much on the library (but I didn't search much either so it could be me!)

**Comment by lminiero** [comment]
Closing as too much changed in the SIP plugin. In case this is still happening with master as well, feel free to open a new issue. If you could check whether the new SIPre plugin (PR available) suffers from the same or not that would help.
